### ISSUE TITLE
IO tests depend too heavily on timing / sleeping

### ISSUE THREAD

**Issue by wolfgangwalther** [PostgREST/postgrest#3424]
I just ran the IO tests on a heavily overloaded system.. and had 32 tests failing, because of some timings related issues. Those pass just fine without load. This is annoying, because I can't continue working on PostgREST while I have other heavy tasks (rebuilding the world for nixpkgs...) running. This is also potentially an issue on much slower systems.

**Comment by steve-chavez** [comment]
Maybe we can separate those tests (`test_io_timing.py`), so they're not always ran. Like what we do with `postgrest-test-memory`.

**Comment by steve-chavez** [comment]
Hm, perhaps some of the pool timeout errors could be solved by using the new `/metrics` endpoint instead of checking the logs.

**Comment by steve-chavez** [comment]
Additionally I think the acq timeout can be less than a second https://hackage.haskell.org/package/hasql-pool-1.1/docs/Hasql-Pool-Config.html#v:acquisitionTimeout.. but not sure if our config allows it now

**Comment by steve-chavez** [comment]
I've also noted sometimes the io tests get stuck: https://github.com/PostgREST/postgrest/actions/runs/8857677646/job/24325423347?pr=3383

I had that happen just one time locally after https://github.com/PostgREST/postgrest/pull/3229. That should be solved by using metrics, I think.

**Comment by steve-chavez** [comment]
io test got stuck again https://github.com/PostgREST/postgrest/actions/runs/8857677646/job/24325423347?pr=3383

We do some waiting when reading the logs https://github.com/PostgREST/postgrest/blob/7e5fd317e6b6060e31e7120837cec95ee4028869/test/io/postgrest.py#L60-L70

But it's capped at 1 second.. it's weird why it gets stuck

---

This looks to be the test that fails
https://github.com/PostgREST/postgrest/blob/7e5fd317e6b6060e31e7120837cec95ee4028869/test/io/test_io.py#L1361-L1389

**Comment by steve-chavez** [comment]
@wolfgangwalther Great idea! I agree.

**Comment by steve-chavez** [comment]
@wolfgangwalther Looking into this now. Admin API has another bug too (connection drop not detected), so will fix that first.

**Comment by steve-chavez** [comment]
> - 200 when ready
> - 503 Service Unavailable while PostgREST is in the process of doing something: During connection to the database, during config reload, during schema cache reload.
> - 500 Internal Server Error as soon as one of the above fails.

Looking more closely at the above proposal, 500 will only happen when the API socket is unreachable. For all other cases we always retry and the status will be 503. This also implies that the `/live` endpoint will now have to reply with a 500 status to be consistent. Thus, this will be a breaking change.

**Comment by steve-chavez** [comment]
> during config reload

Not sure if we should conflate 503 for config reloading. During config reloading we're operational unlike schema cache reloading (makes requests wait) and connection reloading (makes requests fail).

**Comment by steve-chavez** [comment]
> Thus, this will be a breaking change.

To avoid the breaking change and also the problem with responding with 503 while we're operational. How about adding a custom health check endpoint? There we could have the more fine-grained behavior as is.

I'm thinking about a `/loaded` endpoint. Shouldn't this be better since this endpoint will be only for development/testing purposes too?

@wolfgangwalther WDYT?

**Comment by steve-chavez** [comment]
> Do we not have any kind of back-off there? Are we really doing and endless loop of immediately retrying again?

Yes, we do have a backoff. 

>  503 is considered a "temporary" problem, so I would argue that this is actually a bugfix for /live. 
> ...So I don't think we need to treat this as a breaking change.

Agree.

> TLDR: I think we don't need to change /ready and/or /live for the purpose of IO tests. If you still want to do the improvement regarding "schema cache reload in progress" - cool.

Cool. Will finish that on https://github.com/PostgREST/postgrest/pull/3490.

> At this stage it might be easier to just add better log parsing to the IO tests, because all this information (config reload started, config reload finished) etc. are available through the log, I think. This should still allow us to remove the quite random "wait for X seconds" cases I mentioned above.

Parsing the log is still prone to failure though, implementing a `/loaded` would be easier. I also wonder if there's another way to signal we're done reloading state...

**Comment by steve-chavez** [comment]
So check this out, kubernetes has [individual health checks](https://kubernetes.io/docs/reference/using-api/health-checks/#individual-health-checks).

Since the admin server already has `/schema_cache` and `/config`. How about adding a `/schema_cache/ready` and `/config/ready`?

**Comment by steve-chavez** [comment]
> These individual health checks should not be consumed by machines but can be helpful for a human operator to debug a system

https://kubernetes.io/docs/reference/using-api/health-checks/#individual-health-checks

I don't see why we cannot adopt that and have `/ready/schema_cache` and `/ready/config`, it looks like a great idea/design.

> What kinds of failures do you have in mind?

The error message can change, reading the log can hang https://github.com/PostgREST/postgrest/pull/3465 (and this PR didn't solve it, it still happens https://github.com/PostgREST/postgrest/actions/runs/8972240697/job/24639793119).

**Comment by steve-chavez** [comment]
> TLDR: I think we don't need to change /ready and/or /live for the purpose of IO tests. If you still want to do the improvement regarding "schema cache reload in progress" - cool.

Merged that on https://github.com/PostgREST/postgrest/pull/3490.

I tried replacing the `def sleep_until_postgrest_scache_reload()` but it's used in some places where actually it's waiting for the connection recovery to be done, not the schema cache. I'm thinking we not only need a `/ready/config` but also a `/ready/connection`, to make testing independent of the timings.

Overall this part of the code is too hard to refactor/modify because we don't have tests https://github.com/PostgREST/postgrest/issues/1766 (my latest attempt at making progress was on https://github.com/PostgREST/postgrest/issues/1766#issuecomment-2098571024). It's effectively at the "make it work" stage for many years now.

**Comment by steve-chavez** [comment]
> This indicates that there is no backoff involved and instead the reload of the schema cache is tried 10x / s:

Right, the backoff is only for the connection recovery (`establishConnection`), if that succeeds then `work` is repeated and it goes straight ahead to `loadSchemaCache`:

https://github.com/PostgREST/postgrest/blob/1b584f7e9ceda69b956e0b50146d8ba05f2f337b/src/PostgREST/AppState.hs#L365-L398

I remember I kept it that way because the schema cache concept is tied to connections too (they have their own cache https://github.com/PostgREST/postgrest/issues/2620). But it could be improved.

If we really want to improve this part we need the testing utilities (https://github.com/PostgREST/postgrest/issues/1766), otherwise any attempt at refactoring will cause regressions.

**Comment by steve-chavez** [comment]
```
> 08/May/2024:21:06:46 +0200: Attempting to connect to the database...
> 08/May/2024:21:06:46 +0200: Attempting to connect to the database...
> 08/May/2024:21:06:46 +0200: Attempting to connect to the database...
> ...
```

I can reproduce the above by just doing `alter role postgrest_test_authenticator set statement_timeout to '10'
`. To add a way to test the failure, I think we could add a metric `pgrst_schema_cache_failed_loads_total` and check if it increases rapidly.

Additionally, since the connection recovery process is so tied to the schema cache, I think I can use the same backoff for both queries (pg version for the connection, schema cache idem). Otherwise if we have a dedicated backoff for the schema cache query we wouldn't know if it's because the connection is down or if there was a statement timeout.

**Comment by steve-chavez** [comment]
From the log lines, that failure looks like the request was hitting the API server instead of the Admin server.

```
::1 - postgrest_test_anonymous [11/May/2024:18:35:17 +0000] "GET /ready HTTP/1.1" 404 - "" "python-requests/2.31.0"
11/May/2024:18:35:17 +0000: Connection ec7abfbd-8046-4c73-b9b7-33c131add6cd is used
11/May/2024:18:35:17 +0000: Connection ec7abfbd-8046-4c73-b9b7-33c131add6cd is available
```

Because:

- The admin server doesn't log anything when its endpoints are hit.
- A pool connection is used

**Comment by wolfgangwalther** [comment]
The `test_jwt_errors` test fails very often lately, especially in the v12 branch (where I disabled all CI by now), but now it starts to appear on the main branch as well: https://github.com/PostgREST/postgrest/actions/runs/16315210299/job/46079384370

**Comment by taimoorzaeem** [comment]
> The `test_jwt_errors` test fails very often lately, especially in the v12 branch (where I disabled all CI by now), but now it starts to appear on the main branch as well: https://github.com/PostgREST/postgrest/actions/runs/16315210299/job/46079384370

This happens too often for me locally too. I tried debugging this flakiness thinking it maybe related to a slight difference in POSIX seconds between haskell and python, but it isn't the case. I think for now a temporary solution would be to change the `-31` to a `-32` or a higher negative value. This is will prevent CI failures temporarily.

**Comment by steve-chavez** [comment]
>  I think for now a temporary solution would be to change the -31 to a -32 or a higher negative value.

Sounds good to me. Can you add that? @taimoorzaeem
